### ISSUE TITLE
How do I get module locals?

### ISSUE THREAD

**Issue by granmoe** [dsherret/ts-morph#561]
In ts-ast-viewer.com, the SourceFile node has a property called "locals" that contains all symbols in module scope:

<img width="1792" alt="screen shot 2019-03-03 at 12 35 47 pm" src="https://user-images.githubusercontent.com/5620706/53699744-46edb500-3db1-11e9-8d4b-fbb65b17dac6.png">

Is there a way to get this using ts-morph? Or perhaps it is exposed somehow in the normal typescript compiler API?

I'm sorry if there is an obvious place to find the answer. I've searched the typescript docs and ts-morph docs and some of the source code and can't find anything.

Thank you for this incredible library!

**Comment by dsherret** [comment]
Hey @granmoe, always feel free to ask questions... there's a lot that's not really obvious and still a lot not implemented.

The `locals` property is on any node and internal to the compiler api. See [here](https://github.com/Microsoft/TypeScript/blob/8794ebdff5c00a92bb197a5afba02d573681538e/src/compiler/types.ts#L620). Having access to that symbol table would be incredibly useful though. Currently there's [this issue](https://github.com/Microsoft/TypeScript/issues/23719) open on typescript's issue tracker for exposing this.

Right now, you could access this with the following code, but it's not so great:

```ts
import { ts } from "ts-morph";

// ...etc...

project.getProgram().compilerObject; // hack to force binding to happen

const locals = (sourceFile.compilerNode as any)["locals"] as ts.SymbolTable | undefined;

// work with locals here... note that it's compiler api symbols
```

I'm wondering how risky it is to expose this internal property... I'll look into it more later.

Sidenote: I have a pending issue in ts-ast-viewer for toggling on an off internal properties (https://github.com/dsherret/ts-ast-viewer/issues/31).

**Comment by granmoe** [comment]
Thank you for the excellent answer! The links were very helpful.

I simply (heh) need to find all symbols available in module-scope, and I've tried many, many different approaches, but all seem to break down in some way or another. It seems that I'd either have to handle each and every statement in module scope with all of its inherent complexity, or, hopefully, find something *like* `locals` (but hopefully something officially supported and exposed) that has already done the work for me.

This is all in the effort of pursuing my dream project. This is an idea I would devote many years to happily, but I keep having to start over with these crucial, first few steps 😕I am going to keep trying!

**Comment by granmoe** [comment]
I tried getting `locals` using the solution above, but it's always `undefined`. Here's what I'm trying:

```ts
import { ts, Project, Node } from 'ts-morph'

test('Find all identifiers in module-scope (ts-morph, ts compiler locals hack', () => {
  const project = new Project()
  const sourceFile = project.addExistingSourceFile('sample.ts')

  project.getProgram().compilerObject
  const locals = (sourceFile.compilerNode as any)['locals']

  console.log({ locals })
})
```

I also tried some different ways of adding a file to the project with no luck:

```ts
import { ts, Project, Node } from 'ts-morph'

test('Find all identifiers in module-scope (ts-morph, ts compiler locals hack', () => {
  const project = new Project()
  const fileText = 'enum MyEnum {\n}\n'
  const sourceFile = project.createSourceFile('path/to/myNewFile.ts', fileText)

  project.getProgram().compilerObject
  const locals = (sourceFile.compilerNode as any)['locals']

  console.log({ locals })
})
```

Any ideas?

**Comment by dsherret** [comment]
@granmoe sorry, I didn't test out what I said last time. This seems to work:

```ts
import { Project, ts } from "ts-morph";
const project = new Project();
const fileText = 'enum MyEnum {\n}\n';
const sourceFile = project.createSourceFile('path/to/myNewFile.ts', fileText);

sourceFile.getSymbol(); // force binding
const locals = (sourceFile.compilerNode as any)['locals'] as ts.SymbolTable;

console.log({ locals });
```

**Comment by granmoe** [comment]
Awesome! Thank you again 🙂

**Comment by dsherret** [comment]
I'm going to reopen this because I'd like to expose it in the library (even though it's internal to the compiler).

**Comment by dsherret** [comment]
Added `Node#getLocals()`, `Node#getLocalByName(name: string)`, and `Node#getLocalByNameOrThrow(name: string)` in version 3.0.0 (unreleased).

Obviously this relies on a symbol table that's internal in the compiler so use these methods at your own risk knowing they could break more easily than the other methods (added a warning to the jsdoc).

**Comment by dsherret** [comment]
This is now released in 3.0.0. Sorry about the delay! I should have done smaller and faster releases.

Please review [breaking-changes.md](https://github.com/dsherret/ts-morph/blob/master/breaking-changes.md) when upgrading.
